### PR TITLE
Promote odoc errors in dune

### PR DESCRIPTION
**Pull request by rgrinberg**
@aantron it seems like the way dune is generating the index file isn't correct anymore.

### CODE REVIEW

File: test/blackbox-tests/test-cases/github717-odoc-index/index.mld [hunk 1]
```diff
@@ -1 +1 @@
-Test index.
\ No newline at end of file
+{1 Test index}
```
Comment (aantron) [inline test/blackbox-tests/test-cases/github717-odoc-index/index.mld]: If this is a title, it should probably have level 0.

File: test/blackbox-tests/test-cases/odoc-unique-mlds/run.t [hunk 2]
```diff
@@ -8,6 +8,8 @@ Duplicate mld's in the same scope
           odoc _doc/_odoc/lib/root.lib2/root_lib2.odoc
           odoc _doc/_html/root/Root_lib2/.dune-keep,_doc/_html/root/Root_lib2/index.html
           odoc _doc/_odoc/pkg/root/page-index.odoc
+  File "../../../_mlds/root/index.mld", line 3, characters 0-21:
+  '{2': heading level should be lower than top heading level '2'.
```
Comment (aantron) [inline test/blackbox-tests/test-cases/odoc-unique-mlds/run.t]: Where is the code that generates this `.mld` file?

**Review by aantron — COMMENTED** [review]
This:

https://github.com/ocaml/dune/blob/bc77e6dcef1e12dcdded336cf6bc32681f2bbc26/src/odoc.ml#L424

probably needs to be changed to `{0` as well.

File: test/blackbox-tests/test-cases/odoc-unique-mlds/run.t (hunk 2 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/odoc-unique-mlds/run.t]: You probably answered your own question, but it's here:

https://github.com/rgrinberg/jbuilder/blob/fix-odoc-errors/src/odoc.ml#L418-L442

File: test/blackbox-tests/test-cases/github717-odoc-index/index.mld (hunk 1 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/github717-odoc-index/index.mld]: done

**Comment by rgrinberg** [comment]
@aantron I've addressed your comments. I still get these warnings:

```
'{0': heading level should be lower than top heading level '0'.
```

So I think the generated index is still off.

File: src/odoc.ml [hunk 3]
```diff
@@ -421,7 +421,7 @@ module Gen (S : sig val sctx : SC.t end) = struct
     |> List.sort ~compare:(fun (x, _) (y, _) ->
       Lib_name.compare (Lib.name x) (Lib.name y))
     |> List.iter ~f:(fun (lib, modules) ->
-      Printf.bprintf b "{2 Library %s}\n" (Lib_name.to_string (Lib.name lib));
+      Printf.bprintf b "{0 Library %s}\n" (Lib_name.to_string (Lib.name lib));
```
Comment (aantron) [inline src/odoc.ml]: Ah, sorry, I missed the `List.iter` above. The page needs a separate title, with `{0`, which I think Dune is not generating at all right now. Below that, all these section headings should be `{1`.

Comment (rgrinberg) [reply src/odoc.ml]: Done. See the last commit.

**Comment by rgrinberg** [comment]
I'm wondering if this fix deserves a 1.8.2 release. It's a bit of an ugly warning to have for every generated index.mld

**Comment by aantron** [comment]
Looks good.

I would say a quick release is worthwhile, if it is easy to do.

**Pull request merged**